homeassistant/components/dyson/fan.py

```diff
@@ -238,6 +329,255 @@ def supported_features(self) -> int:
+        self._device.turn_off()
+
+    def set_dyson_speed(self, speed: str = None) -> None:
+        """Set the exact speed of the purecool fan."""
+        from libpurecool.const import FanSpeed
+
+        _LOGGER.debug("Set exact speed for fan %s", self.name)
+
+        fan_speed = FanSpeed('{0:04d}'.format(int(speed)))
+        self._device.set_fan_speed(fan_speed)
+
+    def oscillate(self, oscillating: bool) -> None:
+        """Turn on/off oscillating."""
+        _LOGGER.debug("Turn oscillation %s for device %s", oscillating,
+                      self.name)
+
+        if oscillating:
+            self._device.enable_oscillation()
+        else:
+            self._device.disable_oscillation()
+
+    def set_night_mode(self, night_mode: bool) -> None:
+        """Turn on/off night mode."""
+        _LOGGER.debug("Turn night mode %s for device %s", night_mode,
+                      self.name)
+
+        if night_mode:
+            self._device.enable_night_mode()
+        else:
+            self._device.disable_night_mode()
+
+    def set_auto_mode(self, auto_mode: bool) -> None:
+        """Turn auto mode on/off."""
+        _LOGGER.debug("Turn auto mode %s for device %s", auto_mode,
+                      self.name)
+        if auto_mode:
+            self._device.enable_auto_mode()
+        else:
+            self._device.disable_auto_mode()
+
+    def set_angle(self, angle_low: int, angle_high: int) -> None:
+        """Set device angle."""
+        _LOGGER.debug("set low %s and high angle %s for device %s",
+                      angle_low, angle_high, self.name)
+        self._device.enable_oscillation(angle_low, angle_high)
+
+    def set_flow_direction_front(self,
+                                 flow_direction_front: bool) -> None:
+        """Set frontal airflow direction."""
+        _LOGGER.debug("Set frontal flow direction to %s for device %s",
+                      flow_direction_front,
+                      self.name)
+
+        if flow_direction_front:
+            self._device.enable_frontal_direction()
+        else:
+            self._device.disable_frontal_direction()
+
+    def set_timer(self, timer) -> None:
+        """Set timer."""
+        _LOGGER.debug("Set timer to %s for device %s", timer,
+                      self.name)
+
+        if timer == 0:
+            self._device.disable_sleep_timer()
+        else:
+            self._device.enable_sleep_timer(timer)
+
+    @property
+    def oscillating(self):
+        """Return the oscillation state."""
+        return self._device.state and self._device.state.oscillation == "OION"
+
+    @property
+    def is_on(self):
+        """Return true if the entity is on."""
+        if self._device.state:
+            return self._device.state.fan_power == "ON"
+        return False
+
+    @property
+    def speed(self):
+        """Return the current speed."""
+        from libpurecool.const import FanSpeed
+
+        low_speeds = [FanSpeed.FAN_SPEED_1.value,
```

Comment:
Make a dictionary instead and map the device speed values as keys to the speed state that we want to return here. That will simplify the logic below.

Reply:
Done.